### PR TITLE
Support for Single Node Clusters (issue 411)

### LINKED ISSUE #411: [FEATURE] Single Node cluster
Dear all,

I would ask any help in order to add a single node cluster on azure databricks using
- terraform: v0.13.4
- azurerm: 2.32.0
- databrickslab tf provider: 0.2.7

Here my code

```
resource "databricks_cluster" "single_node_etl" {
  cluster_name            = "SINGLENODE-ETL"
  spark_version           = "6.6.x-scala2.11"
  node_type_id            = "Standard_DS3_v2"
  driver_node_type_id    = "Standard_DS3_v2"
  autotermination_minutes = 15
  num_workers = 0
  
  spark_conf = {
    "spark.master" = "local[*]"        # Single Node
    "spark.databricks.cluster.profile" = "singleNode",     # Single Node
    "spark.databricks.passthrough.enabled" = true,             # AAD Passthrough
  }
}
```
but everytime i run terraform apply i receive the error:

`Error: Missing required field: size`

I have tried different configurations but nothing appears to help me.

Any help willl be really appreciated.

Thx

### PR DESCRIPTION
**Pull request by alexott**
this should fix #411 

azcli status
---
 * [ ] **TestAccClusterResource_CreateClusterWithLibraries (222.85s) <= has to be fixed**
 * [ ] ~TestAccDatabricksDBFSFile_CreateViaSource~ (0.57s) <= can be ignored
 * [ ] ~TestAccDatabricksPermissionsResourceFullLifecycle (0.80s)~ <= can be ignored
 * [ ] `TestAccDeleteInvalidMountFails` (11.86s) <= could you test if we can make mounts from single node clusters?...
 * [ ] ~TestAccGroupMemberResource (17.73s)~
 * [ ] ~TestAccSecretAclResourceDefaultPrincipal (63.33s)~
 * [ ] ~TestAccSecretResource (18.47s)~
 * [ ] ~TestAccServicePrincipalResource (10.71s)~
 * [ ] ~TestAccTokenResource (7.59s)~
 * [x] TestAzureAccClusterResource_CreateClusterViaInstancePool (267.38s)
 * [x] access.TestAccPermissionsClusterPolicy (21.53s)
 * [x] access.TestAccPermissionsClusters (297.11s)
 * [x] access.TestAccPermissionsInstancePool (16.17s)
 * [x] access.TestAccPermissionsJobs (17.79s)
 * [x] access.TestAccPermissionsNotebooks (22.75s)
 * [x] access.TestAccPermissionsTokens (14.98s)
 * [x] access/acceptance.TestAccInitialManagePrincipals (3.04s)
 * [x] access/acceptance.TestAccInitialManagePrincipalsGroup (2.53s)
 * [x] access/acceptance.TestAccSecretAclResource (27.52s)
 * [x] access/acceptance.TestAccSecretScopeResource (22.91s)
 * [x] compute.TestAccContext (55.76s)
 * [x] compute.TestAccInstancePools (5.62s)
 * [x] compute.TestAccLibraryCreate (68.14s)
 * [x] compute.TestAccListClustersIntegration (262.62s)
 * [x] compute.TestAzureAccNodeTypes (2.77s)
 * [x] compute/acceptance.TestAccClusterPolicyResourceFullLifecycle (34.33s)
 * [x] compute/acceptance.TestAccJobResource (10.51s)
 * [x] identity.TestAccCreateToken (4.30s)
 * [x] identity.TestAccCreateUser (11.61s)
 * [x] identity.TestAccFilterGroup (4.76s)
 * [x] identity.TestAccGroup (26.48s)
 * [x] identity.TestAccReadUser (4.26s)
 * [x] identity/acceptance.TestAccGroupResource (35.01s)
 * [x] identity/acceptance.TestAccGroupResource_verify_entitlements (25.96s)
 * [x] identity/acceptance.TestAccUserResource (26.98s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace (5.60s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace/Cluster_Policies (0.46s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace/Cluster_Policies_Delete (0.50s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace/Clusters (0.23s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace/DBFS_Files (0.63s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace/Groups (0.69s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace/Instance_Pools (0.24s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace/Jobs (0.21s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace/Notebooks (0.23s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace/Secret_ACLs (0.22s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace/Secret_Scopes (0.16s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace/Secrets (0.26s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace/Tokens (1.11s)
 * [x] internal/sanity.TestAccMissingResourcesInWorkspace/Users (0.66s)
 * [x] internal/sanity.TestAccMutiworkspaceUsedFromNormalMode (6.62s)
 * [x] storage.TestAccCreateFile (15.12s)
 * [x] storage.TestAccInvalidSecretScopeFails (6.15s)
 * [x] storage.TestAccSourceOnInvalidMountFails (33.68s)
 * [x] storage/acceptance.TestAccDatabricksDBFSFile_CreateViaContent (29.04s)
 * [x] workspace/acceptance.TestAccNotebookResourceScalability (24.50s)
 * [x] workspace/acceptance.TestAccWorkspaceConfFullLifecycle (11.47s)

### CODE REVIEW

File: compute/resource_job.go [hunk 1]
```diff
@@ -96,6 +97,15 @@ var jobSchema = internal.StructToSchema(JobSettings{},
 			"`new_cluster` for greater reliability."
 		s["new_cluster"].Description = "Same set of parameters as for " +
 			"[databricks_cluster](cluster.md) resource."
+		p, err := internal.SchemaPath(s, "new_cluster", "num_workers")
+		if err == nil {
```
Comment (nfx) [inline compute/resource_job.go:101]:
```
if p, err := ...; err == nil {
   //..
}
```

**Comment by nfx** [comment]
@alexott could you attach the output of `test.out` file with test report from `make test-azcli` or `make test-azsp`?

**Comment by alexott** [comment]
Running the tests again (this will take some time) - there are failures that aren't directly related to this PR...

**Comment by alexott** [comment]
[out.log](https://github.com/databrickslabs/terraform-provider-databricks/files/5765478/out.log)
[test-report.log](https://github.com/databrickslabs/terraform-provider-databricks/files/5765479/test-report.log)

**Comment by alexott** [comment]
One thing is that with current design, the `num_workers` is always in the JSON payload, even if `autoscale` is specified.  But `autoscale` always wins.  This could be reworked into declaring the `ClusterNumWorkers` as pointer to `int32` and using `omitempty`, but then when we'll need to write everywhere:

```go
numWorkers := int32(X)
Cluster{NumWorkers: &numWorkers....}
```
@nfx WDYT?

File: compute/acceptance/job_test.go [hunk 2]
```diff
@@ -17,6 +17,73 @@ import (
...
+			NumWorkers:   2,
+			SparkVersion: "6.4.x-scala2.11",
+			SparkConf:    nil,
+			AwsAttributes: &AwsAttributes{
+				Availability: "ON_DEMAND",
+			},
+			NodeTypeID: "r3.xlarge",
+		},
+		NotebookTask: &NotebookTask{
+			NotebookPath: "/Users/sri.tikkireddy@databricks.com/demo-terraform/demo-notebook",
```
Comment (nfx) [inline compute/acceptance/job_test.go:39]: replace user-based folders with `/tf-test`. storing emails in the code is not really GDPR-compliant :)

File: compute/acceptance/job_test.go [hunk 3]
```diff
@@ -17,6 +17,73 @@ import (
...
+		t.Skip("Acceptance tests skipped unless env 'CLOUD_ENV' is set")
+	}
+
+	client := common.NewClientFromEnvironment()
+	jobsAPI := NewJobsAPI(context.Background(), client)
+
+	jobSettings := JobSettings{
+		NewCluster: &Cluster{
+			NumWorkers:   2,
+			SparkVersion: "6.4.x-scala2.11",
```
Comment (nfx) [inline compute/acceptance/job_test.go:31]: use `CommonRuntimeVersion()`

File: compute/acceptance/job_test.go [hunk 4]
```diff
@@ -17,6 +17,73 @@ import (
...
+			SparkConf:    nil,
+			AwsAttributes: &AwsAttributes{
+				Availability: "ON_DEMAND",
+			},
+			NodeTypeID: "r3.xlarge",
```
Comment (nfx) [inline compute/acceptance/job_test.go:36]: use `CommonInstanceType()`

File: compute/resource_cluster.go [hunk 5]
```diff
@@ -82,17 +86,39 @@ func resourceClusterSchema() map[string]*schema.Schema {
 			Type:     schema.TypeMap,
 			Computed: true,
 		}
+		s["num_workers"] = &schema.Schema{
+			Type:             schema.TypeInt,
+			Optional:         true,
+			Default:          0,
+			ValidateDiagFunc: validation.ToDiagFunc(validation.IntAtLeast(0)),
```
Comment (nfx) [inline compute/resource_cluster.go:93]: either add `// nolint` or replace with new validation framework, otherwise `make lint` fails

File: compute/resource_job.go [hunk 6]
```diff
@@ -144,6 +154,12 @@ func ResourceJob() *schema.Resource {
 			if err != nil {
 				return err
 			}
+			if js.NewCluster != nil {
+				err = validateClusterDefinition(*js.NewCluster)
+				if err != nil {
```
Comment (nfx) [inline compute/resource_job.go:159]: i'd suggest to collapse it into one line like `if err = validateClusterDefinition(*js.NewCluster); err != nil {`

File: compute/resource_job_test.go [hunk 7]
```diff
@@ -452,3 +385,157 @@ func TestResourceJobDelete_Error(t *testing.T) {
 	qa.AssertErrorStartsWith(t, err, "Internal error happened")
 	assert.Equal(t, "789", d.Id())
 }
+
+func TestResourceJobCreateSingleNode(t *testing.T) {
```
Comment (nfx) [inline compute/resource_job_test.go:389]: could you please also add `TestResourceJobCreateNewClusterFiveWorkers`? It was missing and it's important for this refactoring.

File: compute/resource_job_test.go [hunk 8]
```diff
@@ -452,3 +385,157 @@ func TestResourceJobDelete_Error(t *testing.T) {
...
+	cluster := Cluster{
+		NumWorkers: 0, SparkVersion: "7.3.x-scala2.12", NodeTypeID: "Standard_DS3_v2",
```
Comment (nfx) [inline compute/resource_job_test.go:391]: please run `make fmt`, i think this is incorrect formatting.

File: compute/resource_job_test.go [hunk 9]
```diff
@@ -452,3 +385,157 @@ func TestResourceJobDelete_Error(t *testing.T) {
...
+		Fixtures: []qa.HTTPFixture{
+			{
+				Method:   "POST",
+				Resource: "/api/2.0/jobs/create",
+				ExpectedRequest: JobSettings{
+					NewCluster: &cluster,
+					SparkJarTask: &SparkJarTask{
+						MainClassName: "com.labs.BarMain",
+					},
+					Libraries: []Library{
```
Comment (nfx) [inline compute/resource_job_test.go:407]: please remove `libraries` block, it's tested in previous cases, let's focus these corner case tests on more narrow scope and fill in only required params

File: compute/resource_job_test.go (hunk 7 above)
Comment (nfx) [reply compute/resource_job_test.go:389]: nitpicking: could you move `TestResourceJobCreate*` test to be located right after creation tests? e.g. line 90+.

File: scripts/azcli-integration/main.tf [hunk 10]
```diff
@@ -1,5 +1,10 @@
+terraform {
```
Comment (nfx) [inline scripts/azcli-integration/main.tf:1]: does it work with terraform 0.12? :) plenty of customers still use two-release old version.

**Review by nfx — CHANGES_REQUESTED** [review]

File: compute/resource_cluster.go (hunk 5 above)
Comment (alexott) [reply compute/resource_cluster.go:93]: it was originally failing, but was fixed by bumpting to newer SDK:

```
>make fmt && make lint                13:47 0
✓ Formatting source code with goimports ...
✓ Formatting source code with gofmt ...
✓ Filling vendor folder with library code ...
✓ Linting source code with golangci-lint make sure you run make fmt ...
```

File: compute/resource_job_test.go (hunk 8 above)
Comment (alexott) [reply compute/resource_job_test.go:391]: I did run it...

**Comment by nfx** [comment]
> One thing is that with current design, the `num_workers` is always in the JSON payload, even if `autoscale` is specified. But `autoscale` always wins. This could be reworked into declaring the `ClusterNumWorkers` as pointer to `int32` and using `omitempty`, but then when we'll need to write everywhere:
> 
> ```go
> numWorkers := int32(X)
> Cluster{NumWorkers: &numWorkers....}
> ```
> 
> @nfx WDYT?

it should ideally be handled with `omitempty`. I still don't understand why it doesn't work as expected. @pietern, you were programming GoLang full time for couple of years - do you know the intricacies of Go JSON marshalling and `omitempty` for integers? 

@alexott , there are two options available

1. [Marshaller](https://golang.org/pkg/encoding/json/#Marshaler) and Umarshaller implementation for `compute.Cluster` and `compute.ClusterInfo` (I guess it's more elegant and protected from errors)
2. changing int to int pointer

if you decide to change it to int pointer, besides of changing 17 direct invocation places, you'd need add special handling in:

image

1. StructToSchema - https://github.com/databrickslabs/terraform-provider-databricks/blob/master/internal/reflect_resource.go#L158
2. check if StructToData still works - https://github.com/databrickslabs/terraform-provider-databricks/blob/master/internal/reflect_resource.go#L349
3. DataToStructPointer - https://github.com/databrickslabs/terraform-provider-databricks/blob/master/internal/reflect_resource.go#L380

**Comment by pietern** [comment]
@nfx Looking at the Go tests, `omitempty` on an int field should work as expected, see [encoding/json/encode_test.go](https://github.com/golang/go/blob/846dce9d05f19a1f53465e62a304dea21b99f910/src/encoding/json/encode_test.go#L20-L72). The field `Optionals.Io` will be zero-initialised and is not included in the encoded output.

**Comment by nfx** [comment]
@alexott i trust your years of C++ programming to make the best decision on int pointer vs custom marshaller for those structs :) it's equal amount of code, as far as i see it.

**Comment by alexott** [comment]
@nfx comments are addressed.  I went with custom serialization of JSON for `Cluster` struct (and added more checks into it).
Tested with Terraform 0.12.29, 0.13.5

File: Makefile [hunk 11]
```diff
@@ -40,7 +40,17 @@ install: build
 	@echo "    }"
 	@echo "  }"
 	@echo "}"
-	
+
+install-tf-0.12: build
+	@echo "✓ Installing provider for Terraform 0.12 ..."
+	@mkdir -p '$(HOME)/.terraform.d/plugins/$(shell go version | awk '{print $$4}' | sed 's#/#_#')/'
```
Comment (nfx) [inline Makefile:46]: tf12 doesn't actually need any dir version prefix. look at installation instructions:

```
curl https://raw.githubusercontent.com/databrickslabs/databricks-terraform/master/godownloader-databricks-provider.sh | bash -s -- -b $HOME/.terraform.d/plugins
```

my testing setup is: `$PATH` has 0.12 binary, terraform13 and terraform14 for other version as well.

File: Makefile [hunk 12]
```diff
@@ -40,7 +40,17 @@ install: build
...
+	@cp terraform-provider-databricks '$(HOME)/.terraform.d/plugins/$(shell go version | awk '{print $$4}' | sed 's#/#_#')/terraform-provider-databricks_v$(shell ./terraform-provider-databricks version)'
+	@echo "✓ Use the following configuration for Terraform 0.12 to enable the version you've built"
+	@echo ""
+	@echo 'provider "databricks" {'
+	@echo '  version = "0.3.0"'
```
Comment (nfx) [inline Makefile:51]: version pinning is not really needed for 0.12, 🤷 . it's going against installer script we provide

File: compute/acceptance/cluster_test.go [hunk 13]
```diff
@@ -106,14 +108,18 @@ func (i *instancePoolHCLBuilder) withAwsAttributes(attributesMap map[string]stri
 
 func getCommonLibraries() string {
 	return `
-	library_maven {
-		coordinates = "org.jsoup:jsoup:1.7.2"
-		repo = "https://mavencentral.org"
-		exclusions = ["slf4j:slf4j"]
+	library {
+		maven {
```
Comment (nfx) [inline compute/acceptance/cluster_test.go:112]: thanks for doing this :)

File: compute/clusters.go [hunk 14]
```diff
@@ -2,14 +2,17 @@ package compute
...
 	"fmt"
 	"log"
+	"reflect"
 	"sort"
 	"strings"
 	"time"
 
 	"github.com/databrickslabs/databricks-terraform/common"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/ulule/deepcopier"
```
Comment (nfx) [inline compute/clusters.go:15]: why do we need external dependency? can we do without it?

**Review by nfx — CHANGES_REQUESTED** [review]

File: compute/clusters.go (hunk 14 above)
Comment (alexott) [reply compute/clusters.go:15]: copying of the data using reflection is kind of black magic (for me), I've tried to implement it myself, but gave up after several tries - maybe if I spend more time, I could do it, but not sure that this is a best use of the time

Comment (alexott) [reply compute/clusters.go:15]: Frankly speaking, it works without this custom marshaling - it simply sends both `num_workers` & `autoscale` if both present, and control plane always prefer `autoscale`. I simply afraid that it may stop this way at some point of time, and we'll need to do hotfix release...

File: Makefile (hunk 12 above)
Comment (alexott) [reply Makefile:51]: with current `make install` it doesn't fine provider, unfortunately...

File: compute/clusters.go (hunk 14 above)
Comment (nfx) [reply compute/clusters.go:15]: Single node clusters aren’t major dbu bringers, so I’m okay with that risk. 

Additional libraries just require more processes and adding it just to use in one place might be an overkill :(

File: Makefile (hunk 12 above)
Comment (nfx) [reply Makefile:51]: Even after removing ~/.terraform.d ?...

Comment (alexott) [reply Makefile:51]: Ok, it works if I do install like our install script, and remove the version.  I'll remove this target completely...

File: compute/clusters.go (hunk 14 above)
Comment (alexott) [reply compute/clusters.go:15]: My main concern around this marshalling - if API stop accept both `num_worker` & `autoscale`, this will affect everything, not just single node. If you're ok, I'll revert that commit.

**Review by nfx — APPROVED** [review]
Will merge after running acceptance tests locally :)
